src/components/DropZone/DropZone.tsx

```diff
@@ -111,460 +96,325 @@ export interface DropZoneProps {
-      numFiles: prev.numFiles + acceptedFiles.length,
-    }));
+      if (event.target && dragTargets.current.indexOf(event.target) === -1) {
+        dragTargets.current.push(event.target);
+      }
 
-    if (onDrop) {
-      onDrop(files as File[], acceptedFiles, rejectedFiles);
-    }
+      if (dragging) return;
 
-    if (onDropAccepted && acceptedFiles.length) {
-      onDropAccepted(acceptedFiles);
-    }
+      const {rejectedFiles} = getValidatedFiles(fileList);
 
-    if (onDropRejected && rejectedFiles.length) {
-      onDropRejected(rejectedFiles);
-    }
-    (event.target as HTMLInputElement).value = '';
-  };
+      setDragging(true);
+      setInternalError(rejectedFiles.length > 0);
 
-  private handleDragEnter = (event: DragEvent) => {
-    event.preventDefault();
-    event.stopPropagation();
+      onDragEnter && onDragEnter();
+    },
+    [
+      allowMultiple,
+      disabled,
+      dragging,
+      getValidatedFiles,
+      numFiles,
+      onDragEnter,
+    ],
+  );
 
-    const {dragging, numFiles} = this.state;
-    const {disabled, onDragEnter, allowMultiple} = this.props;
+  const handleDragOver = useCallback(
+    (event: DragEvent) => {
+      stopEvent(event);
+      if (disabled || (!allowMultiple && numFiles > 0)) return;
+      onDragOver && onDragOver();
+    },
+    [allowMultiple, disabled, numFiles, onDragOver],
+  );
 
-    if (disabled || (!allowMultiple && numFiles > 0)) {
-      return;
-    }
+  const handleDragLeave = useCallback(
+    (event: DragEvent) => {
+      event.preventDefault();
 
-    const fileList = getDataTransferFiles(event);
+      if (disabled || (!allowMultiple && numFiles > 0)) return;
 
-    if (event.target && this.dragTargets.indexOf(event.target) === -1) {
-      this.dragTargets.push(event.target);
-    }
+      dragTargets.current = dragTargets.current.filter((el: Node) => {
+        const compareNode =
+          dropNode && 'current' in dropNode ? dropNode.current : document;
+        return el !== event.target && compareNode && compareNode.contains(el);
+      });
 
-    if (dragging) {
-      return false;
-    }
+      if (dragTargets.current.length > 0) return;
 
-    const {rejectedFiles} = this.getValidatedFiles(fileList);
+      setDragging(false);
+      setInternalError(false);
 
-    this.setState({dragging: true, error: rejectedFiles.length > 0});
+      onDragLeave && onDragLeave();
+    },
+    [allowMultiple, disabled, dropNode, numFiles, onDragLeave],
+  );
 
-    if (onDragEnter) {
-      onDragEnter();
-    }
+  useEventListener(dropNode, 'drop', handleDrop);
+  useEventListener(dropNode, 'dragover', handleDragOver);
+  useEventListener(dropNode, 'dragenter', handleDragEnter);
+  useEventListener(dropNode, 'dragleave', handleDragLeave);
+  useEventListener(window, 'resize', adjustSize);
+
+  useComponentDidMount(() => {
+    adjustSize();
+    openFileDialog && triggerFileDialog();
+  });
+
+  const currentID = useUniqueId('DropZone', id);
```

Comment:
We solved this slightly differently in [Checkbox](https://github.com/Shopify/polaris-react/blob/cd2cb606d3b048f74b7f1ae212dbfbe2702e8f9e/src/components/OptionList/components/Checkbox/Checkbox.tsx#L21) in a way that avoids you having to do the currentID/id renaming dance